src/modules/tools/endstops/Endstops.cpp

```diff
@@ -651,11 +657,47 @@ void Endstops::home(axis_bitmap_t a)
 
     this->axis_to_home= a;
 
+    THEROBOT->disable_segmentation = true; // we must disable segmentation as this won't work with it enabled
+
+    // If "release_first" is enabled, start moving the axes away from the endstops until the switches are released
+    this->status = MOVING_FROM_ENDSTOP_FAST;
```

Comment:
This seems to set the state to the new state regardless of if the new setting is enabled or not. I do not like changes to logic that are this drastic, and seem incorrect to me for existing behavior.

Reply:
Though I don't think these transient mode changes had any side effects I changed it as you wish.

Reply:
You have a comment there that does not reflect reality as far as I can see, there is no check for release_first that I can see. Again it could be the diffs are unclear as to what actually changed and where.

Reply:
Thanks, @wolfmanjm 

I don't see any diff artefacts from here, including whitespace:
https://github.com/Smoothieware/Smoothieware/pull/1423/files

According to my browsing (Visual Studio) any of the Homing STATES (`MOVING_TO_ENDSTOP_FAST` etc,) are only evaluated inside `Endstops::read_endstops()`. 

And there, any effect is guarded by `STEPPER[m]->is_moving()`. 
https://github.com/Smoothieware/Smoothieware/blob/32a9f9b3f74df40aaac7cc92122325ea7b97746f/src/modules/tools/endstops/Endstops.cpp#L590-L594

So the transient homing state change (that I had removed now) was irrelevant if no axis is actually moved, even if the ISR were to be called in between. 

An axis is only moved if the flag is set. The overall move is swallowed if `do_release` is never set:
https://github.com/Smoothieware/Smoothieware/blob/32a9f9b3f74df40aaac7cc92122325ea7b97746f/src/modules/tools/endstops/Endstops.cpp#L667-L671

I don't pretend to have the same depth of insight into this. I'm looking forward to seeing your more elegant solution. 

Thanks again for looking into this.

_Mark

Reply:
Yea I ran it through another diff engine, I was confused initially as to why there was a whole new section of code just for a backoff. It seems ok as a first attempt now I see what actually changed.
However I think the same thing can be accomplished with far less changes, I think adding a whole section of code just to back off is probably not the cleanest way to do it, especially as we have a back off section already. So for now I'll leave this here for people to test, and if I have tome to do a different approach I'l note it here. - Thank you for your submission.

Reply:
>I think adding a whole section of code just to back off is probably not the cleanest way to do it, especially as we have a back off section already. 

Yes, I see your point. There is just a fundamental difference in that the existing backoff code is "blind" to the switches whereas the new "release move" is terminated by the (inverted) switches. 

But the existing code could surely be refactored: the existing fast and slow homing moves as well as the new release move are clearly related and could be refactored into a separate function with feedrate and switch polarity parametrized. 

I just wanted to keep the change straight-forward, leaving as much existing code intact as possible and having new behaviour subject to the new flag being set. But I agree that a more "wholehearted" approch would be superior. 

_Mark